sdk/metric/aggregation/aggregation.go

```diff
@@ -0,0 +1,111 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.17
+// +build go1.17
+
+// Package aggregation contains configuration types that define the
+// aggregation operation used to summarizes recorded measurements.
+package aggregation // import "go.opentelemetry.io/otel/sdk/metric/aggregation"
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Aggregation is the aggregation used to summarize recorded measurements.
+type Aggregation struct {
+	// Operation is the kind of operation performed by the aggregation and the
+	// configuration for that operation. This can be Drop, Sum, LastValue, or
+	// ExplicitBucketHistogram.
+	Operation operation
+}
+
+var errAgg = errors.New("aggregation")
+
+// Err returns an error if Aggregation a is invalid, nil otherwise.
+func (a Aggregation) Err() error {
+	if a.Operation == nil {
+		return fmt.Errorf("%w: unset operation", errAgg)
+	}
+	switch v := a.Operation.(type) {
+	case Drop, Sum, LastValue, ExplicitBucketHistogram:
+		if err := v.err(); err != nil {
+			return fmt.Errorf("%w: %v", errAgg, err)
+		}
+		return nil
+	}
+	return fmt.Errorf("%w: unknown %T", errAgg, a.Operation)
+}
+
+// operation is an aggregation operation. The OTel specification does not
+// allow user-defined aggregations, therefore, this is not exported.
+type operation interface {
+	// err returns an error for misconfigured operations.
+	err() error
+}
+
+// Drop drops all data recorded.
+type Drop struct{} // The Drop operation has no parameters.
+
+func (Drop) err() error { return nil }
+
+// Sum summarizes a set of measurements as their arithmetic sum.
+type Sum struct{} // The Sum operation has no parameters.
+
+func (Sum) err() error { return nil }
+
+// LastValues summarizes a set of measurements as the last one made.
+type LastValue struct{} // The LastValue operation has no parameters.
+
+func (LastValue) err() error { return nil }
+
+// ExplicitBucketHistogram summarizes a set of measurements as an histogram
+// with explicitly defined buckets.
+type ExplicitBucketHistogram struct {
+	// Boundaries are the increasing bucket boundary values. Boundary values
+	// define bucket upper bounds. Buckets are exclusive of their lower
+	// boundary and inclusive of their upper bound (except at positive
+	// infinity). A measurement is defined to fall into the greatest-numbered
+	// bucket with a boundary that is greater than or equal to the
+	// measurement. As an example, boundaries defined as:
+	//
+	// []float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 1000}
+	//
+	// Will define these buckets:
+	//
+	// (-∞, 0], (0, 5.0], (5.0, 10.0], (10.0, 25.0], (25.0, 50.0],
+	// (50.0, 75.0], (75.0, 100.0], (100.0, 250.0], (250.0, 500.0],
+	// (500.0, 1000.0], (1000.0, +∞)
+	Boundaries []float64
```

Comment:
I believe this allow users to change boundaries after instantiating a histogram: https://go.dev/play/p/WcrHHk68OaC

Reply:
True, the slice values can be changed here after they are created. I'm open to suggestions for ways to address this.

My initial thoughts are that since this is something the user will set directly, it requires them to understand Go well enough to know that a passed slice can have its values changed if they manipulate the memory values. This is something not unique to this situation and requires the user to have some responsibility in not producing undesired effects.

With that in mind, I can also see how we can aid in preventing this. These structs are meant to act as configuration. I have a prototype for an aggregation implementation in https://github.com/open-telemetry/opentelemetry-go/pull/2954, I was planning to update the creation functions there to accept these structs. We can make sure to copy the boundaries field for the histogram there.

Reply:
Yeah, my concern was more from the "we can't throw unhandled exceptions" angle.  It seems like its probably easier for us to do defensive copying before we validate than to have to handle such cases later.  But if you are planning to do that elsewhere, that should be fine.

Reply:
> Yeah, my concern was more from the "we can't throw unhandled exceptions" angle. It seems like its probably easier for us to do defensive copying before we validate than to have to handle such cases later. But if you are planning to do that elsewhere, that should be fine.

Yes, that was my plan :+1: